### ISSUE TITLE
TSP methods has different functional prescription and function traveling_salesman_problem doesn't count with that

### ISSUE THREAD

**Issue by zabojpetr** [networkx/networkx#5596]
### Current Behavior

The function [traveling_salesman_problem](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L199) choose TSP method by one of function parameters called "method". The problem is, that [christofides](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L127) and [greedy_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1094) has parameters G, weights="weights",... but [simulated_annealing_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1186) and [threshold_accepting_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1404) has parameters G, init_cycle, weights=weights, ... 
Function [traveling_salesman_problem](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L199) call this methods like this [method call](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L319):
```python
best_GG = method(GG, weight)
```
So methods [christofides](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L127) and [greedy_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1094) works well, but [simulated_annealing_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1186) and [threshold_accepting_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1404) ends with error because weights is put to init_cycle. 

### Expected Behavior

Methods should have same functional prescription or function [traveling_salesman_problem](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L199) should has switch for methods.

### Steps to Reproduce

```python
nodes_count = 10
edges = [(u,v) for u in range(nodes_count) for v in range(u+1,nodes_count)]
G = nx.Graph()
G.add_edges_from(edges)

nodes_tsp = [x for x in range(0,nodes_count,3)]
nx.algorithms.approximation.traveling_salesman_problem(G, nodes=nodes_tsp, method=nx.algorithms.approximation.christofides) # works
nx.algorithms.approximation.traveling_salesman_problem(G, nodes=nodes_tsp, method=nx.algorithms.approximation.greedy_tsp) # works
nx.algorithms.approximation.traveling_salesman_problem(G, nodes=nodes_tsp, method=nx.algorithms.approximation.threshold_accepting_tsp) # does not work
nx.algorithms.approximation.traveling_salesman_problem(G, nodes=nodes_tsp, method=nx.algorithms.approximation.simulated_annealing_tsp) # does not work
```

### Environment

Python version: 3.7
NetworkX version: 2.7.1 (but actual version in github will have the same probelm)

**Comment by dschult** [comment]
Thanks very much for this!  

This is an intentional, yet confusing interface and probably should be improved. There are many ways in which this could be addressed. We currently rely on people using the `method` argument by constructing a function with 2 arguments. And our current doc_string lists these two problematic [simulated_annealing_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1186) and [threshold_accepting_tsp](https://github.com/networkx/networkx/blob/7e17ff9007d2e71c2ea2a6d340889c331689343c/networkx/algorithms/approximation/traveling_salesman.py#L1404) functions as valid methods. And they are not currently valid. We even have code for the default method which wraps `threshold_accepting_tsp` so that it takes the required 2 arguments.  So we need to find a fix for this confusion. 

As a workaround while we find a fix, one can use:
`method = lambda G, weight: nx.algorithms.approximation.threshold_accepting_tsp(G, "greedy", weight)` and 
`method = lambda G, weight: nx.algorithms.approximation.simulated_annealing_tsp(G, "greedy", weight)`.  Note that this is the way we current allow people to provide other optional input to the TSP methods. Perhaps better documentation of that process -- including examples, would be helpful too.

**Comment by mjschwenne** [comment]
I thing that there are two main approaches we could take. Either we embrace the lambda expressions as the recommended way to do this, or we could use `**kwarg` or possibly have `traveling_salesman_problem` take in an extra dict which can be passed to which ever method is identified by the `method` keyword. Either way this should be totally optional. Since the definition of the traveling salesman problem only cares about the graph and the edge weights the other parameters like `init_cycle` for `threshold_accepting_tsp` should have a default since these tweak the behavior of the method but do not effect what we are seeking; an approximation of the minimum weight Hamiltonian cycle within a reasonable about of time.

Interestingly, the `init_cycle` that I mentioned says "This argument has no default to make you think about it" directly in the doc string, which does seem a bit odd to me.

**Comment by dschult** [comment]
I think we talked about `init_cycle` last summer too.  It is not clear what the default should be. The option "greedy" seems good at first, but it does take more time than we'd like. An arbitrary list of nodes like `list(G)` might be faster for some cases, but I don't have even experience to know.  I agree that the idea of not providing a default "to make you think about it" is a little strange -- though it is blunt enough to get the message across. It'd just be better if we knew what a good default was.

I think if we found a good default for `init_cycle` we could make the first two arguments to all the tsp methods be `G, weight` and it would simplify the interface. Then putting an example or two in the doc_string where we us a lambda function to supply non-default behavior would be sufficient.

Could we try a couple different `init_cycle` methods on some test graphs to see what tends to work best?

**Comment by mjschwenne** [comment]
Right, testing on some test graphs would probably be the best. It seems like the two methods to consider are 
 
 1. `greedy_tsp`
 2. `list(G) + [next(iter(G))]`
 
These cannot be replaced by say a boolean since, as the doc string mentions, it is possible that we want to use the output of another tsp method as input for the threshold accepting method. 

If some test graphs need to be designed, the principle concern is how large of graphs should we use to explore this problem? It looks like the greedy method is $O(n^2)$ and the other one is linear (I'd assume), so our question is then at what size does the cost of computing the greedy solution become not worth it. I suppose that another, preceding question would be does the greedy tsp solution yield a better approximation when used as the input for `threshold_accepting_tsp`? The number of edges doesn't matter since the tsp problem formulation dictates a complete graph and nothing else which would drastically affect performance comes to mind. 

This is one of those cases where knowing the size of graphs people are interested and capable of using these algorithms in becomes important. I'd guess 10  - 25 vertices, but I have nothing to go on other than my instinct. 

I may try to play with this question in the coming weeks, but my summer job starts tomorrow so we'll have to see how that goes.
